jetson/power_logging/pytorch/layer_execution_time.py

```diff
@@ -0,0 +1,119 @@
+"""Script."""
+
+import torch
+from typing import Any
+import time
+from functools import partial
+import datetime
+import json
+
+
+def load_model(model_name: str) -> Any:
+    """Load model from Pytorch Hub.
+
+    Args:
+        model_name: Name of model.
+            It should be same as that in Pytorch Hub.
+
+    Raises:
+        ValueError: If loading model fails from PyTorch Hub
+
+    Returns:
+        PyTorch model
+    """
+    try:
+        return torch.hub.load('pytorch/vision:v0.10.0', model_name, pretrained=True)
+    except:
+        raise ValueError(
+            f"Model name: {model_name} is most likely incorrect. "
+            "Please refer https://pytorch.org/hub/ to get model name."
+        )
+
+def get_layers(model: torch.nn.Module, name_prefix: str="") -> list[tuple[str, torch.nn.Module]]:
+    """
+    Recursively get all layers in a pytorch model.
+
+    Args:
+        model: the pytorch model to look for layers.
+        name_prefix: Use to identify the parents layer. Defaults to "".
+
+    Returns:
+        a list of tuple containing the layer name and the layer.
+    """
+    children = list(model.named_children())
+
+    if len(children) == 0: # No child
+        result = [(name_prefix, model)]
+    else:
+        # If have children, iterate over each child.
+        result = []
+        for child_name, child in children:
+            # Recursively call get_layers on the child, appending the current
+            # child's name to the name_prefix.
+            layers = get_layers(child, name_prefix + "_" + child_name)
+            result.extend(layers)
+    
+    return result
+
+
+def define_and_register_hooks(model) -> None:
+    """
+    Define and register layer hooks.
+    """
+    layer_time_dict = {}
+
+    for layer_name, layer in get_layers(model):
+        layer.register_forward_pre_hook(partial(layer_time_pre_hook, layer_time_dict, layer_name))
+        layer.register_forward_hook(partial(layer_time_hook, layer_time_dict, layer_name))
+    
+    return layer_time_dict
+
+def layer_time_pre_hook(layer_time_dict, layer_name, module, input) -> None:
+    """
+    Forward pass pre-hook.
+
+    Args:
+        layer_time_dict: dictionary to save hook function output.
+        layer_name: the layer to register hook.
+        module: the module to register hook.
+        input: tuple containing the input arguments to module's forward method.
+    """
+    layer_time_dict[layer_name] = (time.time(), datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"))
```

Comment:
Not enough precision for layer start time, we will need milliseconds too (layers in resnet18 take about 1.5ms to execute). Worth fixing here, to not forget to fix it in the next ticket with Cuda Events